packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/ContextMatcher.java

```diff
@@ -0,0 +1,153 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Value;
+import software.uncharted.terarium.hmiserver.models.dataservice.Grounding;
+import software.uncharted.terarium.hmiserver.models.mira.DKG;
+
+public class ContextMatcher {
+
+	@Value
+	static class CuratedGrounding {
+
+		Map<String, String> identifiers;
+		Map<String, String> context;
+	}
+
+	private static final String CONFIG_FILE = "curated-context.json";
+	private static final Map<String, CuratedGrounding> configData = loadConfig();
+
+	private static Map<String, CuratedGrounding> loadConfig() {
+		try (InputStream inputStream = ContextMatcher.class.getClassLoader().getResourceAsStream(CONFIG_FILE)) {
+			if (inputStream == null) {
+				throw new RuntimeException("Configuration file not found: " + CONFIG_FILE);
+			}
+			ObjectMapper mapper = new ObjectMapper();
+			return mapper.readValue(inputStream, new TypeReference<>() {});
+		} catch (IOException e) {
+			throw new RuntimeException("Failed to load configuration file", e);
+		}
+	}
+
+	@Value
+	static class SearchMatch {
+
+		String key;
+		double score;
+		List<String> matches;
+	}
+
+	/**
+	 * Calculate the Levenshtein distance between two strings
+	 * <a href="https://en.wikipedia.org/wiki/Levenshtein_distance">...</a>
+	 * <a href="https://en.wikipedia.org/wiki/Approximate_string_matching">...</a>
+	 */
+	private static int levenshteinDistance(String a, String b) {
+		int[][] matrix = new int[b.length() + 1][a.length() + 1];
+
+		for (int i = 0; i <= b.length(); i++) {
+			matrix[i][0] = i;
+		}
+		for (int j = 0; j <= a.length(); j++) {
+			matrix[0][j] = j;
+		}
+
+		for (int i = 1; i <= b.length(); i++) {
+			for (int j = 1; j <= a.length(); j++) {
+				if (b.charAt(i - 1) == a.charAt(j - 1)) {
+					matrix[i][j] = matrix[i - 1][j - 1];
+				} else {
+					matrix[i][j] = Math.min(matrix[i - 1][j - 1] + 1, Math.min(matrix[i][j - 1] + 1, matrix[i - 1][j] + 1));
+				}
+			}
+		}
+
+		return matrix[b.length()][a.length()];
+	}
+
+	/**
+	 * Calculate the score of a key based on a single search term.
+	 * @param key The key to compare against.
+	 * @param term The search term.
+	 * @return A SearchMatch object with the key, score, and matched terms.
+	 */
+	private static SearchMatch calculateScore(String key, String term) {
+		String[] keyTerms = key.toLowerCase().split("_");
+		double totalScore = 0;
+		List<String> matches = new ArrayList<>();
+
+		int bestDistance = Integer.MAX_VALUE;
+		boolean matched = false;
+		String termLower = term.toLowerCase();
+
+		for (String keyTerm : keyTerms) {
+			int distance = levenshteinDistance(termLower, keyTerm);
+			if (distance < bestDistance) {
+				bestDistance = distance;
+				if (distance <= Math.min(3, keyTerm.length() / 2)) {
```

Comment:
1. Why are you splitting things by underscores? Doesn't that just make things much more complicated? 
2. You need a comment to explain what this `Math.min(3, keyTerm.length() / 2))` is doing

Reply:
1. I could remove the split, not necessary.
2. I have no idea, I just took a levensthein distance algo online. If you have a better one, I'm happy to update.

Reply:
So I read it again and I think I have a better understanding. The distance logic is a bit weird that it juggles a "distance" and "bestDistance", but that aside ...

It is a threshold to what is considers to be a match, the cost of change has to be smaller than the minimum of
- 3, or
- you changed less than half of yourself

So if you have really short strings like 1-3 characters, this heuristic may be a little wonky.

Reply:
I removed the underscore spliting.
about the math, I tried with just `S` and it returned accurate matches from the list of keys.

Reply:
But if you are worried that the math is wrong, I am more than happy to change it.